### PR TITLE
Bug fix/internal issue #316

### PR DESCRIPTION
**Pull request by gnusi**
Please note that for legal reasons we require you to sign the 
[Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2687/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2692/

File: arangod/IResearch/IResearchViewNode.cpp [hunk 1]
```diff
@@ -285,12 +273,12 @@ int evaluateVolatility(arangodb::iresearch::IResearchViewNode const& node) {
...
-  auto& sortCondition = node.sortCondition();
-  if (!sortCondition.empty() && inInnerLoop) {
+  auto& scorers = node.scorers();
+  if (!scorers.empty() && inInnerLoop) {
     vars.clear();
 
-    for (auto const& sort : sortCondition) {
-      if (::hasDependecies(plan, *sort.node, outVariable, vars)) {
+    for (auto const& scorer : scorers) {
+      if (::hasDependecies(plan, *scorer.node, outVariable, vars)) {
```
Comment (jsteemann) [inline arangod/IResearch/IResearchViewNode.cpp]: should be hasDepende **n** cies here.

File: js/common/modules/@arangodb/aql/explainer.js [hunk 2]
```diff
@@ -1017,7 +1017,22 @@ function processQuery (query, explain) {
         if (node.condition && node.condition.hasOwnProperty('type')) {
           condition = ' ' + keyword('SEARCH') + ' ' + buildExpression(node.condition);
         }
-        return keyword('FOR') + ' ' + variableName(node.outVariable) + ' ' + keyword('IN') + ' ' + view(node.view) + condition + '   ' + annotation('/* view query */');
+
+        var scorers = '';
+        if (node.scorers && node.scorers.length > 0) {
+            scorers = keyword(' LET ');
+            for (var j = 0;;) {
```
Comment (jsteemann) [inline js/common/modules/@arangodb/aql/explainer.js]: could be simplified to
```
scorers = keyword(' LET ' ) + node.scorers.map(function(scorer) {
  return variableName(scorer) + ' = ' + buildExpression(scorer.node);
}).join(', ');
```

**Comment by jsteemann** [comment]
Apart from the above, I am wondering if the change of the scorer serialization data is a compatibility problem, e.g. while performing a rolling upgrade, when some servers in a cluster may use version 3.4 and some already use 3.5/devel.
Coordinators are supposed to be upgraded after all DB servers have been upgraded, so I think at least 3.5 DB servers receiving "old" query plan snippets from 3.4 coordinators need to still handle the old format.

File: arangod/IResearch/AqlHelper.cpp [hunk 3]
```diff
@@ -36,24 +39,307 @@
...
+    case aql::NODE_TYPE_OPERATOR_UNARY_MINUS:
+    case aql::NODE_TYPE_OPERATOR_UNARY_NOT:
+    case aql::NODE_TYPE_OPERATOR_BINARY_AND:
+    case aql::NODE_TYPE_OPERATOR_BINARY_OR:
+    case aql::NODE_TYPE_OPERATOR_BINARY_PLUS:
+    case aql::NODE_TYPE_OPERATOR_BINARY_MINUS:
+    case aql::NODE_TYPE_OPERATOR_BINARY_TIMES:
+    case aql::NODE_TYPE_OPERATOR_BINARY_DIV:
+    case aql::NODE_TYPE_OPERATOR_BINARY_MOD:
+    case aql::NODE_TYPE_OPERATOR_BINARY_EQ:
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/AqlHelper.cpp]: Could you give me an example of how `NODE_TYPE_OPERATOR_BINARY_ARRAY_EQ` can be expressed in AQL?

File: arangod/IResearch/AqlHelper.cpp [hunk 4]
```diff
@@ -36,24 +39,307 @@
...
+    }
+
+    case aql::NODE_TYPE_ATTRIBUTE_ACCESS:
+    case aql::NODE_TYPE_INDEXED_ACCESS:
+    case aql::NODE_TYPE_EXPANSION: {
+      return attributeAccessEqual(lhs, rhs, nullptr);
+    }
+
+    case aql::NODE_TYPE_VALUE:
+    case aql::NODE_TYPE_OBJECT: {
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/AqlHelper.cpp]: good catch, I have to cover object case manually

File: arangod/IResearch/AqlHelper.cpp [hunk 5]
```diff
@@ -36,24 +39,307 @@
...
+    case aql::NODE_TYPE_OPERATOR_UNARY_MINUS:
+    case aql::NODE_TYPE_OPERATOR_UNARY_NOT:
+    case aql::NODE_TYPE_OPERATOR_BINARY_AND:
+    case aql::NODE_TYPE_OPERATOR_BINARY_OR:
+    case aql::NODE_TYPE_OPERATOR_BINARY_PLUS:
+    case aql::NODE_TYPE_OPERATOR_BINARY_MINUS:
+    case aql::NODE_TYPE_OPERATOR_BINARY_TIMES:
+    case aql::NODE_TYPE_OPERATOR_BINARY_DIV:
+    case aql::NODE_TYPE_OPERATOR_BINARY_MOD:
+    case aql::NODE_TYPE_OPERATOR_BINARY_EQ:
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/IResearch/AqlHelper.cpp]: Could you give me an example of how NODE_TYPE_OPERATOR_BINARY_ARRAY_EQ can be expressed in AQL?

File: arangod/IResearch/IResearchViewNode.cpp (hunk 1 above)
Comment (gnusi) [reply arangod/IResearch/IResearchViewNode.cpp]: typo... done

File: js/common/modules/@arangodb/aql/explainer.js (hunk 2 above)
Comment (gnusi) [reply js/common/modules/@arangodb/aql/explainer.js]: done

**Comment by gnusi** [comment]
This PR doesn't touch score serialization at all, the only thing it does is allow users to use scorer function outside of SORT statement, e.g. in arbitrary calculations. That means, while performing a rolling upgrade, those coordinators which are outdated will simply decline unsupported queries even if all db server were already updated. A long story short - these changes are backwards compatible

File: arangod/IResearch/AqlHelper.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/IResearch/AqlHelper.cpp]: NODE_TYPE_OPERATOR_BINARY_EQ: `a == b`
NODE_TYPE_OPERATOR_BINARY_ARRAY_EQ: `[1,2,3] ALL IN b`

File: arangod/IResearch/AqlHelper.cpp [hunk 6]
```diff
@@ -36,24 +39,307 @@
...
+
+  if ((!lhs && rhs) || (lhs && !rhs)) {
+    return false;
+  }
+
+  if (lhs->type != rhs->type) {
+    return false;
+  }
+
+  size_t const n = lhs->numMembers();
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/IResearch/AqlHelper.cpp]: An AstNode has a `flags` member, which contains some computed information.
To my knowledge, that information can always be recomputed, so it should not matter when comparing two AstNodes.

File: arangod/IResearch/AqlHelper.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/IResearch/AqlHelper.cpp]: see above

**Pull request merged**